src/main/kotlin/com/manimdsl/ManimParserVisitor.kt

```diff
@@ -7,9 +7,66 @@ import com.manimdsl.frontend.*
 class ManimParserVisitor : ManimParserBaseVisitor<ASTNode>() {
     val symbolTable = SymbolTableVisitor()
     private val semanticAnalyser = SemanticAnalysis()
+    private var inFunction: Boolean = false
+    private var functionReturnType: Type = VoidType
 
     override fun visitProgram(ctx: ManimParser.ProgramContext): ProgramNode {
-        return ProgramNode(ctx.stat().map { visit(it) as StatementNode })
+        return ProgramNode(
+                ctx.function().map { visit(it) as FunctionNode },
+                ctx.stat().map { visit(it) as StatementNode })
+    }
+
+    override fun visitFunction(ctx: ManimParser.FunctionContext): FunctionNode {
+        inFunction = true
+        val identifier = ctx.IDENT().symbol.text
+        semanticAnalyser.redeclaredVariableCheck(symbolTable, identifier, ctx)
+
+        val type = if (ctx.type() != null) {
+            visit(ctx.type()) as Type
+        } else {
+            VoidType
+        }
+        functionReturnType = type
```

Comment:
Could you explain what the purpose of the global variable `functionReturnType` is? I'm not sure why both `functionReturnType` and `type` are needed here

Reply:
`functionReturnType` is just the way I used to store the function return type so whenever we visit a return statement we can check whether it's compatible with the type of the function. I'm not sure if there is a better way to store the type I need to check each return statement with...

Reply:
Okay, well if people have ideas for removing the global variable we can maybe change it. Could we just remove the local `type` variable then? Just have

```kotlin
functionReturnType = if (ctx.type() != null) {
          visit(ctx.type()) as Type
} else {
          VoidType
}
```

Reply:
Or actually, use `type` everywhere within this function and only assign values to `functionReturnType` for later use?